### PR TITLE
Introduce CLI commands to troubleshoot connectivity issues to the etcd kvstore and clustermesh control plane

### LINKED ISSUE #30937: CFP: Simplify troubleshooting connectivity issues towards control plane components
## Cilium Feature Proposal
        
**Is your proposed feature related to a problem?**

Troubleshooting Cilium agent connection issues towards control plane components (e.g., kube-apiserver, external etcd if Cilium runs in kvstore mode, remote clusters in clustermesh) is typically problematic, as there can be multiple underlying causes, and client libraries often don't surface the exact reason of the failure. 

For instance, `cilium status` currently reports:

```
ClusterMesh:            0/1 clusters ready, 0 global-services
   cluster1: not-ready, 0 nodes, 0 endpoints, 0 identities, 0 services, 0 failures (last: never)
   └  Waiting for initial connection to be established
   └  remote configuration: expected=unknown, retrieved=unknown
   └  synchronization status: nodes=false, endpoints=false, identities=false, services=false
```

if the agent cannot connect to a remote cluster, providing no suggestions to the user about the next troubleshooting steps. Similarly, no information is provided by log messages unless etcd client debug messages are enabled through the `ETCD_CLIENT_DEBUG` environment variable.

**Describe the feature you'd like**

Summarizing the outcome of an offline discussion with @aanm, we propose to introduce a dedicated cilium-dbg command which can be used to run a suite of automated tests (e.g., L3 connectivity, TLS certificates, ...) to assess the connectivity to the given control plane component, and outputs the results in a user-friendly manner. `cilium-dbg` is the optimal component as it is run in the same host-network namespace as the agent itself, and has access to the same configurations and credentials. The same functionality should be provided as a subcommand of the clustermesh-apiserver (and possibly of the cilium-operator) binary itself, as kvstoremesh also connects to the etcd instances in remote clusters.

### PR DESCRIPTION
**Pull request by giorio94**
Troubleshooting etcd connectivity issues, regardless of whether to the Cilium kvstore or to a remote cluster, is a complex activity, as issues can concern network connectivity, TLS certificates mismatch, authn/authz policies and so on.

As an effort to simplify this process, let's introduce new CLI commands part of the Cilium agent, clustermesh-apiserver and kvstoremesh responsible for performing a set of sanity checks, and outputting the result in a user-friendly way:

* `cilium troubleshoot kvstore`:  troubleshoot connectivity towards the Cilium etcd kvstore;
* `cilium troubleshoot clustermesh [clusters...]`:  troubleshoot connectivity towards remote clusters (the check can be optionally limited to the specified cluster names);
* `kubectl exec -it -n kube-system deploy/clustermesh-apiserver -c apiserver -- clustermesh-apiserver clustermesh-dbg troubleshoot`: troubleshoot the clutermesh-apiserver connectivity to the local etcd kvstore;
* `kubectl exec -it -n kube-system deploy/clustermesh-apiserver -c kvstoremesh -- clustermesh-apiserver kvstoremesh-dbg troubleshoot [--include-local] [clusters...]`: troubleshoot the kvstoremesh connectivity to the local etcd kvstore and remote clusters (the check can be optionally limited to the specified cluster names);

Example output of `cilium troubleshoot clustermesh` in case of success is:

```
Found 1 remote cluster configurations

Remote cluster "cheerful-rhino":
📄 Configuration path: /var/lib/cilium/clustermesh/clustermesh1

🔌 Endpoints:
   - https://clustermesh-apiserver.kube-system.svc:2379
     ✅ Hostname resolved to: 172.20.2.218
     ✅ TCP connection successfully established to 172.20.2.218:2379
     ✅ TLS connection successfully established to 172.20.2.218:2379
     ℹ  Negotiated TLS version: TLS 1.3, ciphersuite TLS_AES_128_GCM_SHA256
     ℹ  Etcd server version: 3.5.13

🔑 Digital certificates:
   ✅ TLS Root CA certificates:
      - Serial number:       c0:ae:dd:fc:cd:20:b9:ca:c6:e6:be:ca:8e:b7:a1:03
        Subject:             CN=Cilium CA
        Issuer:              CN=Cilium CA
        Validity:
          Not before:  2024-05-03 07:15:23 +0000 UTC
          Not after:   2027-05-03 07:15:23 +0000 UTC
   ✅ TLS client certificates:
      - Serial number:       37:b3:23:16:b4:a0:61:b1:59:54:59:7a:e6:e2:54:13
        Subject:             CN=remote
        Issuer:              CN=Cilium CA
        Validity:
          Not before:  2024-05-03 07:15:29 +0000 UTC
          Not after:   2027-05-03 07:15:29 +0000 UTC

⚙ Etcd client:
   ✅ Etcd connection successfully established
   ℹ  Etcd cluster ID: 85e29469f06c55e1
```

Failure examples (providing only the relevant snippet for brevity) include:

* The hostname doesn't resolve to any IP:

```
🔌 Endpoints:
   - https://bar.mesh.cilium.io:32380
     ❌ Cannot resolve hostname: lookup bar.mesh.cilium.io: no such host
     ❌ Cannot establish TCP connection to bar.mesh.cilium.io:32380: dial tcp: lookup bar.mesh.cilium.io: no such host
```

* Failed to establish TCP connection:
```
🔌 Endpoints:
   - https://clustermesh2.mesh.cilium.io:9999
     ✅ Hostname resolved to: 172.19.0.4
     ❌ Cannot establish TCP connection to clustermesh2.mesh.cilium.io:9999: dial tcp 172.19.0.4:9999: connect: connection refused
```

* The server certificate is not valid for the given IP/hostname:
```
🔌 Endpoints:
   - https://172.19.0.4:32380
     ✅ TCP connection successfully established to 172.19.0.4:32380
     ❌ Cannot establish TLS connection to 172.19.0.4:32380: tls: failed to verify certificate: x509: certificate is valid for 127.0.0.1, ::1, not 172.19.0.4
  - https://foo.cilium.io:32380
     ✅ Hostname resolved to: 172.19.0.4
     ✅ TCP connection successfully established to 172.19.0.4:32380
     ❌ Cannot establish TLS connection to foo.cilium.io:32380: tls: failed to verify certificate: x509: certificate is valid for clustermesh-apiserver.cilium.io, *.mesh.cilium.io, clustermesh-apiserver.kube-system.svc, not foo.cilium.io
```

* The server certificate is signed by an unknown root CA:
```
🔌 Endpoints:
   - https://clustermesh2.mesh.cilium.io:32380
     ✅ Hostname resolved to: 172.19.0.4
     ✅ TCP connection successfully established to 172.19.0.4:32380
     ❌ Cannot establish TLS connection to clustermesh2.mesh.cilium.io:32380: tls: failed to verify certificate: x509: certificate signed by unknown authority
```

* The client certificate is rejected by the peer as signed by an unknown CA:
```
🔌 Endpoints:
   - https://clustermesh2.mesh.cilium.io:32380
     ✅ Hostname resolved to: 172.19.0.4
     ✅ TCP connection successfully established to 172.19.0.4:32380
     ✅ TLS connection successfully established to 172.19.0.4:32380
     ℹ  Negotiated TLS version: TLS 1.3, ciphersuite TLS_AES_128_GCM_SHA256
     ❌ TLS client authentication failed: remote error: tls: unknown certificate authority

🔑 Digital certificates:
   ...
   ✅ TLS client certificates:
      - Serial number:       37:b3:23:16:b4:a0:61:b1:59:54:59:7a:e6:e2:54:13
        Subject:             CN=remote
        Issuer:              CN=Cilium CA
        Validity:
          Not before:  2024-05-03 07:15:29 +0000 UTC
          Not after:   2027-05-03 07:15:29 +0000 UTC
        ⚠ Cannot verify certificate with the configured root CAs
```

* The etcd user is not authorized to retrieve the data:

```
⚙ Etcd client:
   ❌ Failed to retrieve key from etcd: etcdserver: permission denied
```

I've marked this PR for backport to all stable branches as it qualifies as `Debug tool improvements`, and it doesn't introduce risks considering that it includes CLI changes only.

Documentation changes and the collection of clustermesh-apiserver information as part of the sysdump will be addressed in follow-up PRs.

Fixes: #30937

```release-note
Introduce CLI commands to troubleshoot connectivity issues to the etcd kvstore and clustermesh control plane
```

### CODE REVIEW

File: pkg/clustermesh/common/config.go [hunk 1]
```diff
@@ -177,3 +177,22 @@ func (cdw *configDirectoryWatcher) close() {
...
+// shall be used by CLI tools only, as it doesn't handle subsequent updates.
+func ConfigFiles(cfgdir string) (configs map[string]string, err error) {
+	files, err := os.ReadDir(cfgdir)
+	if err != nil {
+		return nil, err
+	}
+
+	configs = make(map[string]string)
+	for _, f := range files {
+		cfgfile := path.Join(cfgdir, f.Name())
```
Comment (aanm) [inline pkg/clustermesh/common/config.go]: I believe this should be `filepath.Join` instead.

**Review by aanm — APPROVED** [review]
LGTM, just one small thing.

File: pkg/clustermesh/common/config.go (hunk 1 above)
Comment (giorio94) [reply pkg/clustermesh/common/config.go]: Ah, good point, thanks. I didn't know about that difference. I've updated the other two occurrences in the same file as well.

File: pkg/kvstore/etcd_debug.go [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+type indentedWriter struct {
+	w      io.Writer
+	indent []byte
+}
+
+func newIndentedWriter(w io.Writer, indent int) *indentedWriter {
+	return &indentedWriter{w: w, indent: []byte(strings.Repeat(" ", indent))}
+}
+
+func (iw *indentedWriter) NewLine() { iw.w.Write([]byte("\n")) }
```
Comment (derailed) [inline pkg/kvstore/etcd_debug.go:388]: nit: Is there a reason for these to be exported?

**Review by derailed — APPROVED** [review]
@giorio94 Very cool. Nice work Marco!

File: pkg/kvstore/etcd_debug.go (hunk 2 above)
Comment (giorio94) [reply pkg/kvstore/etcd_debug.go:388]: Not really. The idea was just to explicit that these are the methods expected to be used by the consumers of `indentedWriter` (although there aren't others at the moment).

**Review by thorn3r — APPROVED** [review]
This is awesome, great work!

**Pull request merged**